apps/web/src/app/[locale]/challenge/_components/sheet-content-custom.tsx

```diff
@@ -0,0 +1,115 @@
+import { SheetHeader, SheetTitle } from '@repo/ui/components/sheet';
+import Link from 'next/link';
+import { ChevronRight } from '@repo/ui/icons';
+import { TrackChallenge } from '../../tracks/_components/track-challenge-card';
+import { TrackProgress } from '../../tracks/_components/track-progress';
+import { SelectDropdown } from './select-dropdown';
+import { useChallengeRouteData } from '~/app/challenge-route-data.hook';
+import { SORT_KEYS, useProblemExplorerContext } from '~/app/problem-explorer.hooks';
+import { useAllChallengesContext } from '~/app/all-challenges.hook';
+import { useLocalStorage } from '~/utils/useLocalStorage';
+
+export function ExplorerPanel() {
+  const [, setTrackNameStorage] = useLocalStorage('trackName', 'popular');
+  const { getTrack, setTrack, title, setTitle, sortKey, setSortKey } = useProblemExplorerContext();
+  const { allChallenges } = useAllChallengesContext();
+
+  const handleValueChange = (value: string) => {
+    if (value === 'popular') {
+      setTitle('Recommended Challenges');
+      setTrack(allChallenges.popularChallenges);
+      setTrackNameStorage('popular');
+    }
+    if (value === 'beginner') {
+      setTitle('Great for Beginners');
+      setTrack(allChallenges.beginnerChallenges);
+      setTrackNameStorage('beginner');
+    }
+    if (value === 'easy') {
+      setTitle('Great for Learners');
+      setTrack(allChallenges.easyChallenges);
+      setTrackNameStorage('easy');
+    }
+    if (value === 'medium') {
+      setTitle('Great for Enthusiasts');
+      setTrack(allChallenges.mediumChallenges);
+      setTrackNameStorage('medium');
+    }
+    if (value === 'hard') {
+      setTitle('Great for Experts');
+      setTrack(allChallenges.hardChallenges);
+      setTrackNameStorage('hard');
+    }
+    if (value === 'extreme') {
+      setTitle('Great for Masters');
+      setTrack(allChallenges.extremeChallenges);
+      setTrackNameStorage('extreme');
+    }
+    setSortKey(SORT_KEYS.find((sk) => sk.value === value) ?? SORT_KEYS[0]);
+  };
+
+  const { currentChallenge } = useChallengeRouteData();
+
+  const challenges = getTrack;
+  const completedChallenges = challenges
+    .filter((challenge) => {
+      return (
+        challenge.submission.length &&
+        challenge.submission.some((submission) => submission.isSuccessful)
+      );
+    })
+    .map((challenge) => challenge.id);
+
+  const inProgressChallenges = challenges
```

Comment:
These are being calculated on every render, consider wrapping them in a `useMemo`.